Extend actor tests to test `ethusd`

da-kami opened the pull request:
Extend the internals to be able to configure tests per contract symbol.

Note that this results in quite a lot more actor tests, but as far as I could see the test runtime did not go up significantly.

da-kami committed:
Extend actor tests to be test `ethusd`

Extend the internals to be able to configure tests per contract symbol and
enough tests to make sure that sending offers, contract setup, collab and
force close, rollover work.

Co-authored-by: Mariusz Klochowicz <mariusz@klochowicz.com>

klochowicz committed:
refactor(daemon-tests): Restructure OliviaData for different contract symbols

Make it more maintainable in light of adding ETHUSD trading pair.

Whilst at there, remove unused pk fields.

luckysori commented on daemon-tests/src/lib.rs:
```diff
@@ -272,7 +273,7 @@ impl Default for OpenCfdArgs {
             quantity: Contracts::new(100),
             taker_leverage: Leverage::TWO,
-            oracle_data: OliviaData::example_0(),
+            oracle_data: btc_example_0(),
```
🙃 The name `btc_example_0` could apply to so many things.

klochowicz replied:
I thought I improved it over `example_0` 😂 happy to change it over to any other suggestion :)

luckysori commented on daemon-tests/src/maia.rs:
```diff
@@ -27,6 +27,16 @@ pub mod olivia {
         )
     }
+
+    // TODO: Use real-world ETH oracle data.
```
It should be exactly the same as BTCUSD, but with a different event ID. For example:

```
/x/BitMEX/BETH/2022-08-26T01:08:00.price?n=20
```

This means we don't have to wait for olivia to support ETHUSD. We can just pull the values from any random BTCUSD event.

klochowicz replied:
thanks, will use the suggested ID! I wasn't sure what was the format as I wasn't following olivia closely.

klochowicz replied:
I'll stick to the current BTC events for now, and will update with some real values for ETHUSD when olivia supports it (along with different quotes).

luckysori replied:
Can already do this, e.g. https://outcome.observer/h00.ooo/x/BitMEX/BETH/2022-08-29T05:00:00.price?n=20

klochowicz replied:
thanks, I'll post a small follow-up today that uses them! in the meantime, I'll bors this so we can have these running on the CI already.

luckysori commented on daemon-tests/tests/main/collaborative_settlement.rs:
```diff
@@ -2,22 +2,35 @@ use daemon::projection::CfdState;
+    collaboratively_close_an_open_cfd(Position::Long, ContractSymbol::BtcUsd).await;
+}
+
+#[otel_test]
+async fn collaboratively_close_an_open_eth_usd_cfd_maker_going_short() {
+    collaboratively_close_an_open_cfd(Position::Short, ContractSymbol::EthUsd).await;
+}
+
+#[otel_test]
+async fn collaboratively_close_an_open_eth_usd_cfd_maker_going_long() {
+    collaboratively_close_an_open_cfd(Position::Long, ContractSymbol::EthUsd).await;
```
❓ Do we really need long and short versions of these tests? 

I do think there's probably value in having one test per contract symbol because the quote will come from a different place. Although the test won't do anything to check if the values are correct at the end. So even that is debatable.

klochowicz replied:
they don't add anything more, and it's trivial to test it. At least for now I'd keep them whilst we're getting more confidence.

luckysori commented on daemon-tests/tests/main/non_collaborative_settlement.rs:
```diff
@@ -2,24 +2,36 @@ use daemon_tests::open_cfd;
 }
 
-async fn force_close_open_cfd(position_maker: Position) {
+#[otel_test]
+async fn force_close_an_open_eth_usd_cfd_maker_going_short() {
+    force_close_open_cfd(Position::Short, ContractSymbol::EthUsd).await;
+}
+
+#[otel_test]
+async fn force_close_an_open_eth_usd_cfd_maker_going_long() {
+    force_close_open_cfd(Position::Long, ContractSymbol::EthUsd).await;
+}
```
🤔 Also not sure if we need two of these.

luckysori commented on daemon-tests/tests/main/offer.rs:
```diff
@@ -9,7 +9,18 @@ use model::Leverage;
+    taker_receives_offer_from_maker_on_publication(ContractSymbol::EthUsd).await;
+}
+
+// TODO: Potentially create a test where we send out both and ensure we receive as expected
```
👍 I think we should do that as part of this PR.

luckysori commented on daemon-tests/tests/main/order.rs:
```diff
@@ -189,6 +198,9 @@ async fn taker_places_order_for_same_offer_twice_results_in_two_cfds() {
+// be set up and we receive both on the cfd feed
+
 /// To be used for the first contract setup
 async fn first_contract_setup(maker: &mut Maker, taker: &mut Taker, order_id: OrderId) {
```
🔧 What is a "first" contract setup? Every contract setup is the first and only one. Can we rename this?

klochowicz replied:
for some reason Daniel has created 2 functions - one for the "first" and then for "additional" contract setup. I'll investigate whether we need both and remove one if not needed.

I'm surprised by this too.

klochowicz replied:
it turns out that one was enough. incorporated it in a commit in this PR.

luckysori commented on daemon-tests/tests/main/order.rs:
```diff
@@ -189,6 +198,9 @@ async fn taker_places_order_for_same_offer_twice_results_in_two_cfds() {
 }
 
+// TODO: Potentially create a test where we take one ethusd and one btc usd order to ensure both can
+// be set up and we receive both on the cfd feed
```
👎 I don't really see the value of this. It seems like an arbitrary combination that shouldn't tell us anything interesting about the system. I would remove this TODO.

klochowicz replied:
removed :-)

luckysori commented on daemon-tests/tests/main/rollover.rs:
```diff
@@ -36,10 +38,27 @@ async fn rollover_an_open_cfd_maker_going_short() {
+        prepare_rollover(Position::Short, ContractSymbol::EthUsd, eth_example_0()).await;
+
+    // We charge 24 hours for the rollover because that is the fallback strategy if the timestamp of
+    // the settlement-event is already expired
```
🙃 I hate this comment so much :D 

What if we create a method `complete_fee_for_expired_settlement_event` that includes this and why we want to use it (because our test oracle events are hard-coded to be in the past)?

That way we can remove these noisy comments from several places.

luckysori reviewed (approved):
Thanks!

klochowicz committed:
chore(daemon-tests): Remove unneeded first_contract_setup function

There's no need for two functions doing pretty much the same; might as well
assert on both sides during the "first" contract setup.

klochowicz committed:
chore(daemon-tests): Reduce comment noise in rollover tests

Encapsulate comment about particular test setup once instead of repeating it in
every test

klochowicz committed:
feat(deamon-tests): Add a test for offers with different symbols

luckysori commented on daemon-tests/tests/main/offer.rs:
```diff
@@ -34,15 +39,25 @@ async fn taker_receives_offer_from_maker_on_publication(contract_symbol: Contrac
+/// Verify that an offer with a given contract symbol gets published by the
+/// maker an is received by the taker
+async fn assert_offer(maker: &mut Maker, taker: &mut Taker, symbol: ContractSymbol) {
+    publish_offer(maker, symbol).await;
```
❓ Isn't it a bit weird to act and assert in the same test function? I would expect the `publish` bit to happen outside. When reading the tests I was quite confused.

klochowicz replied:
I'll rename the function, as indeed `assert_` name was confusing.

luckysori reviewed (approved):
Thanks!
